artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQMessageProducer.java

```diff
@@ -409,33 +409,42 @@ private void doSendx(ActiveMQDestination destination,
+                        clientSession.createAddress(address, RoutingType.ANYCAST, true);
+                        if (destination.isTemporary()) {
+                           // TODO is it right to use the address for the queue name here?
+                           session.createTemporaryQueue(destination, RoutingType.ANYCAST, address, null, query);
+                        } else {
+                           session.createQueue(destination, RoutingType.ANYCAST, address, null, true, true, query);
+                        }
                      } else {
-                        session.createQueue(destination, RoutingType.ANYCAST, address, null, true, true, query);
+                        throw new InvalidDestinationException("JMSQueue " + address + " cannot be created, autoCreateAddresses is " + query.isAutoCreateAddresses() + " , isAutoCreateQueues=" + query.isAutoCreateQueues());
+                     }
+                  } else {
+                     if (query.isAutoCreateAddresses()) {
+                        clientSession.createAddress(address, RoutingType.MULTICAST, true);
+                     } else {
+                        throw new InvalidDestinationException("JMSTopic " + address + " cannot be created, autoCreateAddresses is " + query.isAutoCreateAddresses());
                      }
-                  } else if (!destination.isQueue() && query.isAutoCreateAddresses()) {
-                     clientSession.createAddress(address, RoutingType.MULTICAST, true);
-                  } else if ((destination.isQueue() && !query.isAutoCreateQueues()) || (!destination.isQueue() && !query.isAutoCreateAddresses())) {
-                     throw new InvalidDestinationException("Destination " + address + " does not exist");
                   }
                } else {
                   if (destination.isQueue()) {
                      ClientSession.QueueQuery queueQuery = clientSession.queueQuery(address);
                      if (!queueQuery.isExists()) {
-                        if (destination.isTemporary()) {
-                           session.createTemporaryQueue(destination, RoutingType.ANYCAST, address, null, query);
+                        if (query.isAutoCreateQueues()) {
+                           if (destination.isTemporary()) {
+                              session.createTemporaryQueue(destination, RoutingType.ANYCAST, address, null, query);
+                           } else {
+                              session.createQueue(destination, RoutingType.ANYCAST, address, null, true, true, query);
+                           }
                         } else {
-                           session.createQueue(destination, RoutingType.ANYCAST, address, null, true, true, query);
+                           throw new InvalidDestinationException("JMSQueue " + address + " cannot be created, address exists but autoCreateQueues is " + query.isAutoCreateQueues());
```

Comment:
I would prefer going with my PR. it's simpler after all if you look closely.

Reply:
for example. another issue on the original code, it's never checking the destination on a defaultDestination.

small issue, but my change also addressed that.


I don't know what you think, but the code I wrote it's easier to read.

Although I understand your concern as this is the JMS layer, and TCK is applied here. But I think it's safe to make the change on this case. I would run the full testsuite,

Reply:
hmmm.. there's another check on createProducer.

It's not worth the effort.. I will just take your version.. and keep the test I'm writing.

Reply:
im just about to send you pr